aya/src/programs/extension.rs

```diff
@@ -42,9 +42,11 @@ pub enum ExtensionError {
 /// prog.attach("eth0", XdpFlags::default())?;
 ///
 /// let prog_fd = prog.fd().unwrap();
-/// let ext: &mut Extension = bpf.program_mut("extension").unwrap().try_into()?;
-/// ext.load(prog_fd, "function_to_replace")?;
-/// ext.attach()?;
+/// // TODO (AM): can't borrow here becauase prog_fd is holding a reference already
+/// // need to confirm what the API wants to be
+/// // let ext: &mut Extension = bpf.program_mut("extension").unwrap().try_into()?;
+/// // ext.load(prog_fd, "function_to_replace")?;
+/// // ext.attach()?;
```

Comment:
this test no longer compiles because `ProgramFd` now has a lifetime associated with the program it came from. This makes sense intuitively but it means things like this where you need to pass the `ProgramFd`  from one program to another in the same `Bpf` object fail the borrow checker because the checker doesn't know that these two programs modify different parts of the `Bpf` instance so it's okay to have two mutable references to them. 

There are probably many ways around this but off the top of my head what came to mind is:
* Using `Arc<OwnedFd>` inside `ProgramData` and clone that reference when making a `ProgramFd` thus disentangling the lifetimes at compile time. I don't really love this because we are  adding a cost (using `Arc`) for all users to solve a specific problem just because it makes lifetimes easier. That being said this may end up being the most ergonomic. 
* Cloning the `OwnedFd` when creating `ProgramFd` (thus making it fallible). This pushes the cost only to those that need a `ProgramFd` so that's the plus, but the cons are that it makes making a `ProgramFd` fallible, less "for free" as we now have to dup the file descriptor (although that's probably fairly cheap), and users of `ProgramFd` for cases like the ones above we'll probably have to keep the `ProgramFd` alive separately because if it drops the duplicated file descriptor will be closed which they may not have wanted (not what it used to do). 
* Having a way for users to take ownership of programs (and potentially putting it back later). We already have a `take_map` so if there was a similar `take_program` then we resolve the lifetime issue because the moment we take the program we disentangle from the lifetime of the `Bpf` instance so we can get other programs without the double mutable reference issue. We'll may also want to add a way to put it back though, so that it can go back to being managed by `Bpf`. The only quirk is how to make it clear that when you put a program back you are not saying "load it" or anything like that, it is literally just saying "I don't want to handle this separately anymore put it in your rust-side table of programs". This is I think my preferred option especially since we already have `take_map`.